Issue #13411: Fix NPE in UnnecessarySemicolonAfterTypeMemberDeclarationCheck for Local Class

relentless-pursuit opened the pull request:
This commit addresses issue #13411, where the UnnecessarySemicolonAfterTypeMemberDeclarationCheck was throwing a NullPointerException when dealing with a local class inside a switch statement.

A null check has been added to the isSemicolon method of the UnnecessarySemicolonAfterTypeMemberDeclarationCheck class to ensure that the method correctly handles null input and prevents the exception.

Resolves: #13411

rnveach commented:
@relentless-pursuit What don't you understand from the CI failures?

romani commented:
>Rule violated for class com.puppycrawl.tools.checkstyle.checks.coding.UnnecessarySemicolonAfterTypeMemberDeclarationCheck: branches covered ratio is 0.96, but expected minimum is 1.00


https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/47611429/job/tvp54fdmwqiwec3t#L1425

All CI must be green, if any violation happens, please review and fix.

You just need to add test to support your changes. Code from issue description must be added to prove that what user reported is covered.

relentless-pursuit commented:
> @relentless-pursuit What don't you understand from the CI failures?

I believe i didn't add test methods. I am working on it.

relentless-pursuit commented:
@romani As I can see, all the CIs passed excpet semaphore CI. However, on observing the logs

> > Rule violated for class com.puppycrawl.tools.checkstyle.checks.coding.UnnecessarySemicolonAfterTypeMemberDeclarationCheck: branches covered ratio is 0.96, but expected minimum is 1.00
> 
> https://ci.appveyor.com/project/Checkstyle/checkstyle/builds/47611429/job/tvp54fdmwqiwec3t#L1425
> 
> All CI must be green, if any violation happens, please review and fix.
> 
> You just need to add test to support your changes. Code from issue description must be added to prove that what user reported is covered.

I reduce the CI failure. Made code changes as per the logs of CI failures. Howerver, two of the CIs are still failing. I skimmed through the failing logs but couldn't find anything substantial related to my code. Do they fall often?

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessarySemicolonAfterTypeMemberDeclarationCheck.java:
```diff
@@ -233,6 +233,16 @@ private void checkEnumConstant(DetailAST ast) {
+     * @param ast the DetailAST node which needs to be checked
+     * @return true if the given ast is semicolon, false otherwise
+     */
+    public static boolean checkProcess(DetailAST ast) {
```
Please use only Input file based testing. As all other tests are doing.

relentless-pursuit replied:
I have been struggling with the folder structure. Right now, there are two input files in the `UnnecessarySemicolonAfterTypeMemberDeclarationCheckTest` i.e. `InputUnnecessarySemicolonAfterTypeMemberDeclaration.java` located in `com/puppycrawl/tools/checkstyle/checks/coding/unnecessarysemicolonaftertypememberdeclaration/InputUnnecessarySemicolonAfterTypeMemberDeclaration.java`

similarly, `InputUnnecessarySemicolonAfterTypeMemberDeclarationRecords.java` located in `src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unnecessarysemicolonaftertypememberdeclaration/InputUnnecessarySemicolonAfterTypeMemberDeclarationRecords.java`

So, when i put my new input file inside any of the sub-directory of unnecessarysemicolonaftertypememberdeclaration it gives me error. I have two questions. Kindly bear me as i am contributing to this repo for the first time. 

Should I write the unit test in UnnecessarySemicolonAfterTypeMemberDeclarationCheckTest or create a different class for it? If i am writing it in UnnecessarySemicolonAfterTypeMemberDeclarationCheckTest, where do i place the unit test as there is very strong check on file name and location. Would something like this work?
```
unnecessarysemicolonaftertypememberdeclaration
 > semicolontests
   > InputIsSemicolonWithNullAst.java
 > InputUnnecessarySemicolonAfterTypeMemberDeclaration.java
```

relentless-pursuit replied:
@romani I have pushed a new patchset. Kindly review.

romani replied:
@relentless-pursuit , no worries, I know you are New. We and Ci will guide you through a process. Just  keep calm, and follow CI and us and we will succeed.

relentless-pursuit replied:
Surely. Thanks for the assurance. And, I am currently going through the CI logs and making changes. I will push a new patchset once i have resolved most CI logs issue to the best of my capabilities. And, ask for further guidance if stuck. Thanks again.

relentless-pursuit replied:
@Romani, finally after so many attempts from saturaday, I could see a green. Can you review and give me your feedback to proceed further?

relentless-pursuit replied:
I tried understanding the failing issue mentioned in 'Check no closed issue references / check_issues'. But, i couldn't decipher. Any input on the same will be helpful.

nrmancuso replied:
You need to remove module and comment at https://github.com/relentless-pursuit/checkstyle/blob/3b6d5b812082ebd5a81ecdfe5de1bbd0c50ae47c/config/projects-to-test/openjdk20-excluded.files#L40

relentless-pursuit replied:
> You need to remove module and comment at https://github.com/relentless-pursuit/checkstyle/blob/3b6d5b812082ebd5a81ecdfe5de1bbd0c50ae47c/config/projects-to-test/openjdk20-excluded.files#L40

I assume the comment must be added by a script. As a beginnner who wants to keep contributing to this repo, how can we avoid it in the first place?

nrmancuso replied:
Do not be afraid to use `git blame` / GitHub history to understand context and details about existing code. This module and corresponding comment were added in the scope of https://github.com/checkstyle/checkstyle/issues/13402

relentless-pursuit replied:
@romani it seems to be passing all CIs. Can you review it once again?

romani reviewed (changes requested):
Items:

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessarysemicolonaftertypememberdeclaration/InputUnnecessarySemicolonAfterTypeMemberDeclarationNullAst.java:
```diff
@@ -0,0 +1,23 @@
+package com.puppycrawl.tools.checkstyle.checks.coding.unnecessarysemicolonaftertypememberdeclaration;
+
+public class InputUnnecessarySemicolonAfterTypeMemberDeclarationNullAst {
+    public static void main(String[] args) { // ok
```
Please remove all ok comments.
I know that some check demands file to have at least one, please add it over class

romani reviewed (changes requested):
Last item

romani commented:
Fix CI 
Please read and follow git flow described at https://checkstyle.org/beginning_development.html#Starting_Development

relentless-pursuit commented:
> Fix CI Please read and follow git flow described at https://checkstyle.org/beginning_development.html#Starting_Development

Thank you. After I updated the java file, I am getting a CI failure. And, going into the logs it says
`groovy: JAVA_HOME is not defined correctly, can not execute: /usr/lib/jvm/adoptopenjdk-11-hotspot-ppc64el/bin/java`

Bit confusing.

romani reviewed (approved):
Ok to merge

relentless-pursuit commented:
> Ok to merge

my bad, give me some time, i was confused over the error. will revert the changes i just pushed now.

relentless-pursuit commented:
> Ok to merge

in the recent push, i have updated the changes as requested by you. the earlier CI failure was confusing me. And, i was trying to revert the changes i made i.e. the comments to see if it was a failure from my end. After your comments on ok to merge, i made my final push for now based on your requested review. I have just added // ok to the beginning of the class.

nrmancuso commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessarysemicolonaftertypememberdeclaration/InputUnnecessarySemicolonAfterTypeMemberDeclarationNullAst.java:
```diff
@@ -0,0 +1,24 @@
+        switch (args.length) {
+            case 0:
+                final int k = 12;
+                class Local {
+                    int j = k;
+                }
```
Please add test cases as follows:
1. Add another local class declaration to this block statement WITH an unnecessary semicolon.
2. Add another local class with fields that have extra semicolons, to show that we still place violations on them. See the check documentation for some examples.

nrmancuso reviewed (changes requested):
Please generate a check regression report as detailed at https://github.com/checkstyle/contribution/tree/master/checkstyle-tester#report-generation using GitHub action.

Item:

relentless-pursuit commented:
I had some issues in understanding the CI log. Hence, closing this PR. I will raise a new PR.

nrmancuso commented:
@relentless-pursuit please do not open and close PRs, outline issues here and we will guide you.

relentless-pursuit commented:
> @relentless-pursuit please do not open and close PRs, outline issues here and we will guide you.

Sure, Do i reopen this PR? And, close the one I created.
